### ISSUE TITLE
Ctc bpe conformer overfit

### ISSUE THREAD

**Issue by Pavloveuge** [NVIDIA/NeMo#7244]
Hi, I'm trying to train a bpe-level conformer and am facing overfitting.

In this regard, I have a few questions and I would be very grateful for the answers to them:
1) Have you encountered overfitting of the BPE conformer and BPE models in general?
2) Have you tried to experiment with char level vs bpe level models? If yes, what are you facing? I mean over/underfitting, stability, approximate increase in metrics.
3) I saw [configs ](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/conformer/conformer_ctc_bpe.yaml) for training conformer from nemo, thanks for posting them, but I couldn't figure out what the global batch size was. [2048](https://github.com/NVIDIA/NeMo/issues/4183#issuecomment-1128444967)?
4) What are some tips for tuning regularization for a conformer?
5) How trainning stability depends on the size of the dictionary?

Before that, I used the char-level conformer, and did not encounter overfitting. I don't have a lot of data, 370 hours for one domain and 130 for another. Therefore, I first train the base model, and then finetune it on domain data. I did some experiments:
1) I finetuned stt_ru_conformer_ctc_large using my domain data and got pretty good quality (better my char level model)
2) I trained my base model, just like I did before with char-level models, and then tried to finetuned it on domain data

The training data configurations are completely the same, except for the weights of models and tokens of tokenizers (the sizes of models and tokenizers are the same).
The length of my audios is from 0.1 to 10 seconds, the size of the bpe dictionary is 128, the average transcription length is 53 characters or 31 tokens.
I use the same (as in nemo models) spec augs, except mask value (mean instead of 0).

I used AdamW, a constant scheduler with lr = 0.0001, global batch size = 512 and faced with overfitting (see plots), then I reduced lr by 10 times and increased global batchs size by 3 times (1536 in total), but this did not solve the problem cardinally. The metrics I get (before the start overfitting) are worse than not only the experiment on finetuning the base model from NeMo, but also worse than the experiments on finetuning my char-level base model.

Plots for char based model:
![newplot (18)](https://github.com/NVIDIA/NeMo/assets/49618087/d32b972b-d27f-4332-9af1-0fc3bb810bc4)
![newplot (15)](https://github.com/NVIDIA/NeMo/assets/49618087/3818469d-2a1c-4b50-a989-75b867915842)

Plots for bpe based model:
![newplot (20)](https://github.com/NVIDIA/NeMo/assets/49618087/6c35247c-fa0b-457e-9675-1222838749e9)
![newplot (19)](https://github.com/NVIDIA/NeMo/assets/49618087/788d546d-b82f-44e4-b015-600df36437df)

@titu1994 

Thank you in advance

**Comment by titu1994** [comment]
Thanks for the detailed comment ! 

1) Yes we definitely encounter overfitting on small datasets. We tackle it with a combination of training longer with lower Peak LR, higher spec augment, larger model trained for shorter time or a combo of these. There's no one fix, and usually the only real fix is to use more data. In data constrained scenarios, training with high LR for shorter number of steps might be beneficial, along with careful tuning of spec augment. 

2) We don't use char Tokenizers for anything other than Asian languages (Chinese, Japanese, Korean, etc). That too only because subwords aren't going to reduce the vocab size nor will they find very common merges. Both char and subword models can overfit, but char models for simpler languages will usually train a bit slower and get slightly worse wer. However there are always exceptions to such findings. 

3) 2048 global batch size for large models, 1024 for small. They will usually be mentioned in the yaml config files. 

4) For conformer there is specifically their dropout values (3 of them), spec Aug, and LR peak as well as decay schedule. We find Cosine schedule does sometimes prevent overfitting too fast, and gets slightly better scores sometimes, but usually Noam schedule with high enough peak LR will be good enough. 

5) Larger the vocabulary, more complicated the alignment task. Note that this is not like crossentropy, simply expanding the vocab size will not make it easier for the model to learn. Alignment losses like CTC and RNNT will struggle beyond certain vocab sizes. You can somewhat circumvent the issue using Sampled Softmax RNNT though. RNNT is also much more robust to vocab sizes as compared to CTC models. We generally stick to 1024 vocab size at max for most languages, but Asian languages can go up to 8k vocab (unique characters only). With sampled softmax, I personally was able to train upto 65K vocab size as a extreme experiment with sampled softmax vocab size of 1024 tokens. Didn't help WER (slightly worse then baseline) but didn't struggle in convergence either. 

For users, we absolutely recommend always start from a pretrained model - no exceptions. Even if the model does not exist in that language, simply take the equivalent english checkpoint and swap out it's decoder for your language and this finetuning will almost always get better results than training from scratch. 

Infact we find that when we ourselves start from a pretrained SSL checkpoint, we can significant reduce the effect of overfitting itself even if we don't use enough data for a XXLarge (1.2 B param model). 

Btw, given the very small number of tokens, it suggests the model is does not want a large vocab size. Try to use 128 at max.

**Comment by titu1994** [comment]
Also one string way to regularize your models is to increase the strength of Fast Emit. We can go up to 0.01 without blowing up our models. 
Though of course, this only works for transducers

**Comment by Pavloveuge** [comment]
Thanks for the detailed and fast response. I still have a few questions about dropout, spec aug and scheduler. Now I use the same dropouts and spec augs as in Nemo. What LR do you recommend for Cosine scheduler?

Perhaps you have experimented with dropouts and spec augs and can you add something about them?

**Comment by Pavloveuge** [comment]
Does it make sense to try InterCTC?
